### PR TITLE
Made simulated test timing more smooth

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Previously the simulated tests used absolute simulated time to determine how long to sleep if it needed to in order to make things appear "real-time". However if things ran slowly for a bit, things later could run extremely fast because they wouldn't sleep at all (because the time was "lagging behind" so much of where we thought we should be).

I changed the system to use relative timings, and also fixed a small bug where we only slept half as long as we needed to. Everything now runs at a much more consistent and realistic speed.

### Testing Done

manual verification

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/simulated_tests/simulated_test_fixture.h [hunk 1]
```diff
@@ -138,15 +138,16 @@ class SimulatedTestFixture : public ::testing::Test
             non_terminating_function_validators);
 
     /**
-     * Puts the current thread to sleep until the difference between the current wall time
-     * and the wall_start_time is >= the difference between current_time and zero. This
-     * "synchronizes" wall time with simulation time by slowing down execution if needed.
+     * Puts the current thread to sleep such that each simulation step will take
+     * the desired about of real-world "wall" time.
```
Comment (kelvinkoon) [inline src/software/simulated_tests/simulated_test_fixture.h]: Looks like this should read "amount"

File: src/software/simulated_tests/simulated_test_fixture.h [hunk 2]
```diff
@@ -138,15 +138,16 @@ class SimulatedTestFixture : public ::testing::Test
...
      *
-     * @param wall_start_time The time at which the simulated test started, in wall time
-     * @param current_time The current time in the simulated test
+     * @param wall_start_time The time at which the most recent simulation step started,
+     * in wall time
+     * @param desired_wall_tick_time How long each simulation step should take
+     * in wall-clock time
```
Comment (kelvinkoon) [inline src/software/simulated_tests/simulated_test_fixture.h:147]: Is there somewhere mentioning what "wall-clock" time is and how it differs from "wall" time?

**Review by kelvinkoon — CHANGES_REQUESTED** [review]
:heavy_check_mark:

File: src/software/simulated_tests/simulated_test_fixture.h (hunk 2 above)
Comment (MathewMacDougall) [reply src/software/simulated_tests/simulated_test_fixture.h:147]: They're the same, just both common ways of saying the same thing (as opposed to saying "real-time")

File: src/software/simulated_tests/simulated_test_fixture.h (hunk 1 above)
Comment (MathewMacDougall) [reply src/software/simulated_tests/simulated_test_fixture.h]: :+1: fixed

File: src/software/simulated_tests/simulated_test_fixture.cpp [hunk 3]
```diff
@@ -146,17 +146,16 @@ void SimulatedTestFixture::updateSensorFusion()
...
+    auto current_tick_wall_time_duration =
         std::chrono::duration_cast<std::chrono::milliseconds>(wall_time_now -
                                                               wall_start_time);
-    auto ms_to_sleep =
-        std::chrono::milliseconds(static_cast<int>(current_time.getMilliseconds())) -
-        wall_time_since_sim_start;
+    auto ms_to_sleep = std::chrono::milliseconds(
+                           static_cast<int>(desired_wall_tick_time.getMilliseconds())) -
+                       current_tick_wall_time_duration;
+    std::cout << "sleeping for " << ms_to_sleep.count() << std::endl;
```
Comment (jonathanlew) [inline src/software/simulated_tests/simulated_test_fixture.cpp]: looks like a debug print out

**Review by jonathanlew — CHANGES_REQUESTED** [review]
just a nit

File: src/software/simulated_tests/simulated_test_fixture.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/software/simulated_tests/simulated_test_fixture.cpp]: whoops, removed

**Review by jonathanlew — APPROVED** [review]

**Review by EvanMorcom — APPROVED** [review]
LGTM :fire: 💯

**Review by kelvinkoon — APPROVED** [review]
LGTM pending merge conflicts.

**Comment by MathewMacDougall** [comment]
merge conflicts resolved

**Review by kelvinkoon — APPROVED** [review]

**Review by EvanMorcom — COMMENTED** [review]
LGTM 👍

**Pull request merged**